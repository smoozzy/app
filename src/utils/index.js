/**
 * Checks condition and throws error
 *
 * @param {boolean|function|number|string|Array|Object} condition
 * @param {string} message
 * @throws {Error} condition must be positive
 */
export function assert(condition, message) {
    if (!condition) {
        throw new Error(`[bootstrap] ${message}`);
    }
}

/**
 * Inheritance
 *
 * You should remember about calling of parent constructor
 *
 * ```javascript
 * function Child() {
 *     Parent.call(this);
 * }
 * ```
 *
 * @param {function} child
 * @param {function} parent
 */
export function extend(child, parent) {
    const proto = Object.create(parent.prototype);

    Object.entries(child.prototype).forEach(([name, prop]) => {
        proto[name] = prop;
    });
    proto.constructor = child;  // override constructor if exists
    child.prototype = proto;

    return child;
}
