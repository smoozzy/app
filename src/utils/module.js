import {assert} from './index';


/**
 * @typedef {Object} AppModule
 * @extends {RouteConfig}
 * @property {string} name - 'module' - used as route name and store module name
 * @property {StoreOptions} [store] - plugs store module on app initialisation
 */


/**
 * Returns component wrapper
 *
 * @param {function|ComponentOptions} component
 * @param {Object} extension
 * @param {ModuleStore} [extension.store]
 * @return {function(): Promise<ComponentOptions>}
 */
function getComponentWrapper(component, extension) {
    return function componentWrapper() {
        const promise = typeof component === 'function'
            ? component()  // dynamic loading
            : Promise.resolve(component);  // static loading

        return promise.then(component => Object.assign(component, extension));
    };
}

/**
 * Normalizes children route names
 *
 * @param {AppModule} module
 * @return {RouteConfig}
 */
function normalizeRouteNames(module) {
    const {
        name,
        children,
        /** @type RouteConfig */ ...route
    } = module;

    if (Array.isArray(children)) {
        // we assume if module has children one will have name

        // find overview (default) page in nested routes
        const overview = children.find(({path}) => !path /* path is empty string or undefined */ || path === '/');

        if (overview === undefined) {
            route.name = name;
        } else {
            // Fix: [vue-router] named route has a default child route (path: '' or '/').
            // When navigating to this named route, the default child route will not be rendered.
            overview.name = name;
        }

        // check children routes
        route.children = children.map(normalizeRouteNames);

    } else if (name !== undefined) {
        // it can be route with redirect only
        route.name = name;
    }

    return route;
}

/**
 * Normalizes app module
 *
 * Converts module config to route config, registers store module and modals
 *
 * @param {AppModule} module
 * @return {RouteConfig}
 */
export function normalize(module) {
    const {
        name,
        component,
        components,
        store,
    } = module;

    const route = normalizeRouteNames(module);

    // extends module with properties like store
    const extension = {};

    if (store !== undefined) {
        // prevent to replace module by module with the same name
        assert(!this.$store.state[name], 'root store should not have module with the same name');
        extension.store = this.$store.registerModuleStore(name, store);
    }

    // loading of vue components
    if (component !== undefined) {
        route.component = getComponentWrapper(component, extension);
    }
    if (components !== undefined) {
        route.components = Object.entries(components).reduce((memo, [name, component]) => {
            memo[name] = getComponentWrapper(component, extension);
            return memo;
        }, {});
    }

    return route;
}
