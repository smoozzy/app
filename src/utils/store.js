import Vuex from 'vuex';
import {extend} from './index';

/* Explanation: why we need this black magic with store constructor
 *
 * In strict mode a store runs watcher that monitors any change from outside.
 * There are only one correct way to change state it is mutation or
 * the `commit` method (mutation internally uses the `commit` method).
 *
 * We try to replace the `commit` method to support tracking of a changes and
 * manage committing flag in the both of stores (root and module) at the same
 * time.
 *
 * Let's look at store constructor. We see initialization of internal structures,
 * overriding of the `commit` method and registration of options (store param)
 * as root module (creating of module collection). Next step is initialization
 * of submodules, mutations and actions. See `installModule` function.
 * (https://github.com/vuejs/vuex/blob/dev/src/store.js)
 *
 * `installModule` function creates local context for wrappers of mutations and
 * actions that uses overridden `commit` method (by Store constructor).
 * It means we can not replace the `commit` method for mutations and actions
 * of root module. Following code will produce error in tests because we try
 * to replace `commit` method after parent did initialization of root module:
 *
 * ```js
 * class MyStore extends Store {
 *     constructor(options) {
 *         super(options);
 *
 *         this.commit = function myCommitMethod() {};
 *     }
 * }
 * ```
 *
 * You can ask why can not use a magic with ES6 class syntax? We have restriction
 * with `super` keyword. See description on MDN.
 * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super#Description)
 *
 * Our implementation try to replace the `commit` method at runtime parent
 * constructor.
 *
 * Enjoy!
 */

const {Store} = Vuex;
const NAMESPACE_DIVIDER = '/';

/* Root store
 */

/**
 * Constructs root store
 *
 * @constructor
 * @extends Store
 * @param {StoreOptions} [options]
 * @return {RootStore}
 */
export function RootStore(options) {
    // saved origin commit method from parent constructor
    let /** @type function */ originalCommit;

    // plugged module stores
    /** @type Object<string,ModuleStore> */
    this._stores = {};

    const that = this;

    // workaround for guard in strict mode (module store)
    function commit(type, payload, options) {
        const path = type.split(NAMESPACE_DIVIDER);
        path.pop();  // remove mutation name

        const module = that._stores[path.join(NAMESPACE_DIVIDER)];
        let committing;

        if (module) {  // set committing status if module exists
            committing = module._committing;
            module._committing = true;
        }

        // run mutation
        originalCommit(type, payload, options);

        if (module) {  // restore committing status
            module._committing = committing;
        }
    }

    // black magic to manage the `commit` method
    Object.defineProperty(this, 'commit', {
        get() {
            // return method from prototype until parent constructor overrides it
            return originalCommit ? commit : Store.prototype.commit;
        },
        set(/** @type function */ value) {
            originalCommit = value;
        }
    });

    // call parent constructor
    Store.call(this, options);
}

RootStore.prototype = {
    /**
     * Creates module store and registers it as module in root store
     *
     * @this RootStore
     * @param {string|string[]} path
     * @param {StoreOptions|Module} module
     * @param {ModuleOptions} [options]
     * @return {ModuleStore}
     */
    registerModuleStore(path, module, options) {
        const {
            namespaced = false,
        } = module;

        // always register module store as namespaced module
        module.namespaced = true;

        // check params and register module
        this.registerModule(path, module, options);
        module.namespaced = namespaced;

        const moduleStore = new ModuleStore(module, this);
        this._stores[typeof path === 'string' ? path : path.join(NAMESPACE_DIVIDER)] = moduleStore;

        return moduleStore;
    },

    /**
     * Unregisters module store
     *
     * @this RootStore
     * @param {string|string[]} path
     */
    unregisterModuleStore(path) {
        const key = typeof path === 'string' ? path : path.join(NAMESPACE_DIVIDER);
        const /** @type ModuleStore */ moduleStore = this._stores[key];

        // check module store
        if (!moduleStore) {
            return false;
        }

        // unlink module store from root store
        moduleStore.unlink();
        this._stores[key] = null;

        this.unregisterModule(path);
    },
};

extend(RootStore, Store);


/* Module store
 */

/**
 * Constructs module store
 *
 * @constructor
 * @extends Store
 * @param {StoreOptions} options
 * @param {RootStore} rootStore
 * @return {ModuleStore}
 */
export function ModuleStore(options, rootStore) {
    // saved origin commit method from parent constructor
    let /** @type function */ originalCommit;

    // workaround for guard in strict mode (module store)
    function commit(type, payload, options) {
        const committing = rootStore._committing;
        rootStore._committing = true;
        originalCommit(type, payload, options);
        rootStore._committing = committing;
    }

    // black magic to manage the `commit` method
    Object.defineProperty(this, 'commit', {
        get() {
            // return method from prototype until parent constructor overrides it
            return originalCommit ? commit : Store.prototype.commit;
        },
        set(/** @type function */ value) {
            originalCommit = this._originalCommit = value;
        }
    });

    // call parent constructor
    Store.call(this, options);
}

ModuleStore.prototype = {
    /**
     * Restores origin commit method and removes relationship with root store
     *
     * @this ModuleStore
     */
    unlink() {
        this.commit = this._originalCommit;
    }
};

extend(ModuleStore, Store);


/* Wrapping of Vuex helper methods
 */

/**
 * Changes context to root component for given function
 *
 * @param {function} fn
 * @return {function(...*): *}
 */
function changeContext(fn) {
    return function setRootContext(...args) {
        return fn.apply(this.$root, args);
    };
}

/**
 * Wraps mapper (vuex helper) to provide options param
 * that allows to switch context to root component
 *
 * @param {function(string, Array|Object): Object<string,function>} mapper
 * @result {function(string, Array|Object, {root: boolean}): Object<string,function>}
 */
function wrapMapper(mapper) {
    return function mapSection(namespace, section, options = {}) {
        const map = mapper(namespace, section);

        const params = typeof namespace === 'string'
            ? options
            : (section || options);  // `options` is default value here

        if (!params.root) {
            return map;
        }

        return Object.entries(map).reduce((memo, [key, fn]) => {
            memo[key] = changeContext(fn);
            return memo;
        }, {});
    };
}

export const mapState = wrapMapper(Vuex.mapState);
export const mapGetters = wrapMapper(Vuex.mapGetters);
export const mapMutations = wrapMapper(Vuex.mapMutations);
export const mapActions = wrapMapper(Vuex.mapActions);
