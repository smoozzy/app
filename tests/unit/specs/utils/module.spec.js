import Vue from 'vue';
import Vuex from 'vuex';
import {normalize} from '@/utils/module';
import {RootStore} from '@/utils/store';

Vue.use(Vuex);

function getRootComponentMock() {
    return {
        $store: new RootStore({
            strict: true,
        }),
    };
}


describe('Module utils', () => {
    it('should fix navigation to default children route', () => {
        // case: no children
        expect(normalize({
            name: 'auth',
            path: '/auth',
        })).toMatchSnapshot();

        // case: children with own paths
        expect(normalize({
            name: 'auth',
            path: '/auth',
            children: [{
                name: 'signin',
                path: 'signin',
            }, {
                name: 'signup',
                path: 'signup',
            }]
        })).toMatchSnapshot();

        // case: child with default route
        expect(normalize({
            name: 'auth',
            path: '/auth',
            children: [{
                name: 'signin',
                path: '',  // default route
            }, {
                name: 'signup',
                path: 'signup',
            }]
        })).toMatchSnapshot();

        // case: routes without name
        expect(normalize({
            path: '*',
            redirect: {
                name: 'page404',
            },
        })).toMatchSnapshot();
    });

    it('should raise error if exists store module with the same name', () => {
        const root = getRootComponentMock();

        root.$store.registerModule('auth', {
            state: {
                module: 'auth-root',
            },
        });

        expect(() => {
            normalize.call(root, {  // setup correct `this`
                name: 'auth',
                path: '/auth',
                store: {
                    state: {
                        module: 'auth',
                    },
                },
            });
        }).toThrow();
    });

    it('should register module store as module in root store', () => {
        const root = getRootComponentMock();

        normalize.call(root, {  // setup correct `this`
            name: 'auth',
            path: '/auth',
            store: {
                state: {
                    module: 'auth',
                },
                getters: {
                    module(state) {
                        return state.module;
                    },
                },
            },
        });

        expect(root.$store.state.auth.module).toBe('auth');
        expect(root.$store.getters['auth/module']).toBe('auth');
    });

    it('should register local store for module component', done => {
        const root = getRootComponentMock();

        const route = normalize.call(root, {  // setup correct `this`
            name: 'math',
            path: '/math',
            component: {
                name: 'Math',
                render: h => h('div', 'Math page'),
            },
            store: {
                state: {
                    counter: 1,
                },
                getters: {
                    square(state) {
                        return Math.pow(state.counter, 2);
                    },
                },
                mutations: {
                    increment(state) {
                        state.counter += 1;
                    },
                },
            },
        });

        route.component().then(component => {
            // root store
            expect(root.$store.state.math.counter).toBe(1);
            expect(root.$store.getters['math/square']).toBe(1);

            // local store
            expect(component.store instanceof Vuex.Store).toBeTruthy();
            expect(component.store.state.counter).toBe(1);
            expect(component.store.getters.square).toBe(1);

            // changing via root store mutation
            root.$store.commit('math/increment')

            expect(root.$store.state.math.counter).toBe(2);
            expect(root.$store.getters['math/square']).toBe(4);

            expect(component.store.state.counter).toBe(2);
            expect(component.store.getters.square).toBe(4);

            // changing via module store mutation
            component.store.commit('increment');

            expect(root.$store.state.math.counter).toBe(3);
            expect(root.$store.getters['math/square']).toBe(9);

            expect(component.store.state.counter).toBe(3);
            expect(component.store.getters.square).toBe(9);

            done();
        });
    });

    it('should register module store for each component for named views', done => {
        const root = getRootComponentMock();

        const route = normalize.call(root, {  // setup correct `this`
            name: 'statistic',
            path: '/stat',
            components: {
                linear: () => Promise.resolve({
                    name: 'LinearRegression',
                    render: h => h('div', 'Linear regression'),
                }),
                polynomial: () => Promise.resolve({
                    name: 'PolynomialRegression',
                    render: h => h('div', 'Polynomial regression'),
                }),
            },
            store: {
                state: {
                    name: 'statistic',
                },
            },
        });

        Promise.resolve()
            .then(() => {
                // root store
                expect(root.$store.state.statistic.name).toBe('statistic');
            })
            .then(() => {
                // linear component store
                return route.components.linear().then(component => {
                    expect(component.store instanceof Vuex.Store).toBeTruthy();
                    expect(component.store.state.name).toBe('statistic');
                });
            })
            .then(() => {
                // polynomial component store
                return route.components.polynomial().then(component => {
                    expect(component.store instanceof Vuex.Store).toBeTruthy();
                    expect(component.store.state.name).toBe('statistic');
                });
            })
            .then(done);
    });
});
