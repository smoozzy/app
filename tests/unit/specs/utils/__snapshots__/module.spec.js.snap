// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module utils should fix navigation to default children route 1`] = `
Object {
  "name": "auth",
  "path": "/auth",
}
`;

exports[`Module utils should fix navigation to default children route 2`] = `
Object {
  "children": Array [
    Object {
      "name": "signin",
      "path": "signin",
    },
    Object {
      "name": "signup",
      "path": "signup",
    },
  ],
  "name": "auth",
  "path": "/auth",
}
`;

exports[`Module utils should fix navigation to default children route 3`] = `
Object {
  "children": Array [
    Object {
      "name": "auth",
      "path": "",
    },
    Object {
      "name": "signup",
      "path": "signup",
    },
  ],
  "path": "/auth",
}
`;

exports[`Module utils should fix navigation to default children route 4`] = `
Object {
  "path": "*",
  "redirect": Object {
    "name": "page404",
  },
}
`;
